hs_tools_resource/utils.py

```diff
@@ -58,3 +62,48 @@ def get_SupportedSharingStatus_choices():
             ['Discoverable', 'Discoverable'],
             ['Private', 'Private'],
             ]
+
+
+def do_work_when_launching_app_as_needed(app_shortkey, res_shortkey, user):
+    """
+    check whether there are extra work needed to be done when the app is launched.
+    :param app_shortkey: shortkey of the app tool resource to be launched
+    :param res_shortkey: shortkey of the resource launching the app tool resource
+    :param user: requesting user
+    :return: error message if an exception is raised; otherwise, return None
+    """
+
+    # When app resource has iRODS federation target path and target resource defined as
+    # extended metadata, the resource needs to be pushed to the specified iRODS federation path
+    # in the specified iRODS storage resource.
+    try:
+        res = get_resource_by_shortkey(res_shortkey)
+        app_res = get_resource_by_shortkey(app_shortkey)
+
+        # check whether irods_path_key and irods_resc_key are added as extended metadata of the
+        # app tool resource, and if they are, pass them out as context variables for app launcher
+        # to push resource to federated iRODS path accordingly
+        irods_path_key = 'irods_federation_target_path'
```

Comment:
I haven't made it through everything, yet but is there a better place we could put these keys?

Reply:
I initially put these keys into settings.py, then I thought they should not be configurable since after we advertise this functionality and web app tool resource users start to use them, we should not change them since they would force users to learn new keys to use and migrate existing web app tool resources to use new keys as well. I do plan to run the names of these additional keys by @dtarb and @danames and @zhiyuli by email later. I don't have strong opinions on this. If you think it is better to put these keys into settings.py, let me know and I'll do so.

Reply:
I agree we should not change them and wasn't thinking of settings.py.  I was just wondering if there might be a place to put them that is more central and visible for app metadata keys.  Somewhere to help keep them documented where these, as well as other app metadata keys can be found.  I'm not sure where that would be or even if we have such a thing setup.

Reply:
I'll create a ```apps_keys.py`` under hs_tools_resource to hold all supported app keys. Let me know if you have other ideas.